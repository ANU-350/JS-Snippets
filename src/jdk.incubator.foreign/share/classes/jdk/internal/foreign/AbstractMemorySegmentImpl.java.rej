diff a/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java b/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java	(rejected hunks)
@@ -136,6 +136,16 @@ public abstract class AbstractMemorySegmentImpl implements MemorySegment, Memory
                 base(), min(), size);
     }
 
+    public void copyFromSwap(MemorySegment src, long elemSize) {
+        AbstractMemorySegmentImpl that = (AbstractMemorySegmentImpl)src;
+        long size = that.byteSize();
+        checkAccess(0, size, false);
+        that.checkAccess(0, size, true);
+        SCOPED_MEMORY_ACCESS.copySwapMemory(scope, that.scope,
+                that.base(), that.min(),
+                base(), min(), size, elemSize);
+    }
+
     private final static VarHandle BYTE_HANDLE = MemoryLayout.ofSequence(MemoryLayouts.JAVA_BYTE)
             .varHandle(byte.class, MemoryLayout.PathElement.sequenceElement());
 
