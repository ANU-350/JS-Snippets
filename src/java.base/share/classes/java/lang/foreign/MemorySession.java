/*
 * Copyright (c) 2021, 2022, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package java.lang.foreign;

import jdk.internal.foreign.MemorySessionImpl;
import jdk.internal.javac.PreviewFeature;
import jdk.internal.ref.CleanerFactory;

/**
 * A memory session models the lifetime of a memory segment.
 * <p>
 * Segments associated with a memory session can only be accessed while the session is {@linkplain #isAlive() alive},
 * and by the {@linkplain #isOwnedBy(Thread) thread} associated with the session (if any).
 * <p>
 * Memory sessions can be <em>unbounded</em> or <em>bounded</em>. An unbounded memory session is obtained by calling
 * {@link MemorySession#global()}. An unbounded memory session is always alive. As a result, the segments associated
 * with an unbounded session are always accessible and their backing regions of memory are never deallocated. Moreover,
 * memory segments associated with unbounded sessions can be accessed from any thread.
 * <p>
 * Conversely, a bounded memory session has a start and an end. Bounded memory sessions can be managed either
 * explicitly, (i.e. using an {@linkplain Arena arena}) or implicitly, by the garbage collector. When a bounded memory
 * session ends, it is no longer {@linkplain #isAlive() alive}, and subsequent operations
 * on the segments associated with that bounded session (e.g. {@link MemorySegment#get(ValueLayout.OfInt, long)})
 * will fail with {@link IllegalStateException}. Moreover, to prevent temporal safety, access to memory segments associated with
 * bounded sessions might be restricted to specific threads (see below).
 *
 * <h2 id = "thread-confinement">Bounded memory sessions and thread-confinement</h2>
 *
 * Bounded memory sessions provide strong temporal safety guarantees: a memory segment associated with a bounded
 * session can only be accessed <em>while</em> the session is alive. The costs associated with maintaining this
 * safety invariant can vary greatly, depending on how many threads have access to the memory segment associated to a
 * memory session. For instance, if a memory session is created and closed by one thread, and the segments associated
 * with it are only ever accessed by that very same thread, ensuring correctness is simple: either the thread accessing
 * the segment finds that the segment session is still alive, in which case access succeeds, or it finds that the segment
 * session is no longer alive, in which case access fails with an exception.
 * <p>
 * Conversely, if a bounded session is associated with segments that can be accessed by multiple threads, or if the
 * session can be closed by a thread other than the accessing thread, the situation is much more complex. For instance,
 * a segment {@linkplain Arena#allocate(long, long) allocated} in an arena, can be accessed <em>while</em> the associated
 * arena is being closed, concurrently, by another thread. Even in this case, memory sessions provide strong temporal
 * safety guarantees, but doing so can incur in a higher performance impact: the Java runtime has to determine that
 * no segment associated with a bounded session is being accessed, before marking a session can be marked as not alive.
 * <p>
 * For this reason, bounded memory sessions can be further divided into two categories: <em>thread-confined</em> memory sessions,
 * and <em>shared</em> memory sessions.
 * <p>
 * Confined memory sessions, support strong thread-confinement guarantees. Upon creation, they are assigned an
 * {@linkplain #isOwnedBy(Thread) owner thread}, typically the thread which initiated the creation operation.
 * After creating a confined memory session, only the owner thread will be allowed to directly manipulate the segments
 * associated with the memory session. Any attempt to perform resource access from a thread other than the
 * owner thread will fail with {@link WrongThreadException}.
 * <p>
 * Shared memory sessions, on the other hand, have no owner thread; as such, segments associated with shared memory sessions
 * can be accessed by multiple threads. This might be useful when multiple threads need to access the same resource concurrently
 * (e.g. in the case of parallel processing).
 *
 * <h2 id="implicit">Implicitly-managed bounded memory sessions</h2>
 *
 * Managing bounded memory session explicitly, using arenas, while powerful, must be used with caution. An arena must always
 * be closed when no longer in use (this is done using {@link Arena#close()}). A failure to do so
 * might result in memory leaks. To mitigate this problem, clients can obtain an {@linkplain MemorySession#implicit() obtain}
 * bounded memory sessions that are managed implicitly, by the garbage collector. These sessions end at some unspecified
 * time <em>after</em> they become <a href="../../../java/lang/ref/package.html#reachability">unreachable</a>, as shown below:
 *
 * {@snippet lang = java:
 * MemorySegment segment = MemorySegment.allocateNative(100, MemorySession.implicit());
 * ...
 * segment = null; // the session might end after this instruction
 *}
 *
 * Bounded sessions that are managed implicitly can be useful to manage long-lived segments, where timely deallocation
 * is not critical, or in unstructured cases where the boundaries of the lifetime associated with a memory session
 * cannot be easily determined. As shown in the example above, a memory session that is managed implicitly cannot end
 * if a program references to one or more segments associated with that session. This means that implicitly managed
 * session can naturally support safe access from multiple threads.
 *
 * @implSpec
 * Implementations of this interface are thread-safe.
 *
 * @see Arena
 * @see MemorySegment
 * @see SymbolLookup
 * @see Linker
 * @see VaList
 *
 * @since 19
 */
@PreviewFeature(feature=PreviewFeature.Feature.FOREIGN)
public sealed interface MemorySession permits MemorySessionImpl {

    /**
     * {@return {@code true}, if this memory session is alive}
     */
    boolean isAlive();

    /**
     * {@return test if the provided thread is the owner thread associated with this memory session}
     * @param thread the thread to be compared against this session's owner thread.
     */
    boolean isOwnedBy(Thread thread);

    /**
     * Runs a critical action while this memory session is kept alive.
     * @param action the action to be run.
     */
    void whileAlive(Runnable action);

    /**
     * Creates a new bounded memory session that is managed, implicitly, by the garbage collector.
     * The returned session can be shared across threads.
     *
     * @return a new bounded memory session that is managed, implicitly, by the garbage collector.
     */
    static MemorySession implicit() {
        return MemorySessionImpl.createImplicit(CleanerFactory.cleaner());
    }

    /**
     * {@return an unbounded memory session}
     */
    static MemorySession global() {
        return MemorySessionImpl.GLOBAL;
    }
}
