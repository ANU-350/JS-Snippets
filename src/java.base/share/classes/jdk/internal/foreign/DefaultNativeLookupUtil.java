package jdk.internal.foreign;

import java.lang.foreign.FunctionDescriptor;
import java.lang.foreign.Linker;
import java.lang.foreign.MemoryLayout;
import java.lang.invoke.MethodHandle;
import java.lang.invoke.MethodHandles;
import java.lang.invoke.MethodType;

public final class DefaultNativeLookupUtil {

    private static final MethodHandle INT_IS_ONE;

    static {
        try {
            INT_IS_ONE = MethodHandles.lookup().findStatic(DefaultNativeLookupUtil.class, "intIsOne", MethodType.methodType(boolean.class, int.class));
        } catch (Exception e) {
            throw new ExceptionInInitializerError(e);
        }
    }

    private DefaultNativeLookupUtil() {}

    public static MethodHandle downcallOfVoid(String name, MemoryLayout... types) {
        return Linker.nativeLinker().downcallHandle(
                Linker.nativeLinker().defaultLookup().find(name).orElseThrow(),
                FunctionDescriptor.ofVoid(types));
    }

    public static MethodHandle downcall(String name, MemoryLayout resLayout, MemoryLayout... types) {
        return Linker.nativeLinker().downcallHandle(
                Linker.nativeLinker().defaultLookup().find(name).orElseThrow(),
                FunctionDescriptor.of(resLayout, types));
    }

    public static MethodHandle intIsOne(MethodHandle original) {
        return MethodHandles.filterReturnValue(original, INT_IS_ONE);
    }

    public static boolean intIsOne(int returnValue) {
        return returnValue == 1;
    }


}
