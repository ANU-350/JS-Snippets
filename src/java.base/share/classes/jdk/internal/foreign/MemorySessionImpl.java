/*
 *  Copyright (c) 2019, 2022, Oracle and/or its affiliates. All rights reserved.
 *  DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 *  This code is free software; you can redistribute it and/or modify it
 *  under the terms of the GNU General Public License version 2 only, as
 *  published by the Free Software Foundation.  Oracle designates this
 *  particular file as subject to the "Classpath" exception as provided
 *  by Oracle in the LICENSE file that accompanied this code.
 *
 *  This code is distributed in the hope that it will be useful, but WITHOUT
 *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 *  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 *  version 2 for more details (a copy is included in the LICENSE file that
 *  accompanied this code).
 *
 *  You should have received a copy of the GNU General Public License version
 *  2 along with this work; if not, write to the Free Software Foundation,
 *  Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 *   Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 *  or visit www.oracle.com if you need additional information or have any
 *  questions.
 *
 */

package jdk.internal.foreign;

import java.lang.foreign.MemorySession;
import java.lang.foreign.SegmentAllocator;
import java.lang.invoke.MethodHandles;
import java.lang.invoke.VarHandle;
import java.lang.ref.Cleaner;
import java.lang.ref.Reference;
import java.util.Objects;

import jdk.internal.misc.ScopedMemoryAccess;
import jdk.internal.vm.annotation.ForceInline;

/**
 * This class manages the temporal bounds associated with a memory segment as well
 * as thread confinement. A session has a liveness bit, which is updated when the session is closed
 * (this operation is triggered by {@link MemorySession#close()}). This bit is consulted prior
 * to memory access (see {@link #checkValidState()}).
 * There are two kinds of memory session: confined memory session and shared memory session.
 * A confined memory session has an associated owner thread that confines some operations to
 * associated owner thread such as {@link #close()} or {@link #checkValidState()}.
 * Shared sessions do not feature an owner thread - meaning their operations can be called, in a racy
 * manner, by multiple threads. To guarantee temporal safety in the presence of concurrent thread,
 * shared sessions use a more sophisticated synchronization mechanism, which guarantees that no concurrent
 * access is possible when a session is being closed (see {@link jdk.internal.misc.ScopedMemoryAccess}).
 */
public non-sealed class MemorySessionImpl implements MemorySession, SegmentAllocator {

    final State state;

    public MemorySessionImpl(State state) {
        this.state = state;
    }

    public State state() {
        return state;
    }

    @Override
    public final boolean isAlive() {
        return state.isAlive();
    }

    @Override
    public boolean isCloseable() {
        return true;
    }

    @Override
    public final Thread ownerThread() {
        return state.ownerThread();
    }

    @Override
    public final void whileAlive(Runnable action) {
        Objects.requireNonNull(action);
        state.acquire();
        try {
            action.run();
        } finally {
            state.release();
        }
    }

    @Override
    public void addCloseAction(Runnable runnable) {
        Objects.requireNonNull(runnable);
        state.checkValidStateWrapException();
        state.resourceList.add(State.ResourceCleanup.ofRunnable(runnable));
    }

    @Override
    public void close() {
        if (isCloseable()) {
            state.close();
        } else {
            throw new UnsupportedOperationException("Cannot close!");
        }
    }

    @Override
    public final MemorySession asNonCloseable() {
        if (!isCloseable()) return this;
        return new MemorySessionImpl(state) {
            @Override
            public boolean isCloseable() {
                return false;
            }
        };
    }

    @ForceInline
    public static void checkValidState(MemorySession session) {
        ((MemorySessionImpl)session).state.checkValidStateWrapException();
    }

    public static void addOrCleanupIfFail(MemorySession session, State.ResourceCleanup cleanup) {
        try {
            session.addCloseAction(cleanup);
        } catch (Throwable ex) {
            cleanup.cleanup();
            throw ex;
        }
    }

    @Override
    public final boolean equals(Object obj) {
        return obj instanceof MemorySessionImpl sessionImpl &&
                sessionImpl.state == state;
    }

    @Override
    public final int hashCode() {
        return state.hashCode();
    }

    public final static MemorySessionImpl GLOBAL = new MemorySessionImpl(new SharedSessionState.OfImplicit()) {

        @Override
        public void addCloseAction(Runnable runnable) {
            Objects.requireNonNull(runnable);
            // do nothing
        }

        @Override
        public boolean isCloseable() {
            return false;
        }
    };

    public final static MemorySessionImpl heapSession(Object o) {
        return createImplicit(o);
    }

    public final static MemorySessionImpl createConfined(Thread thread, Cleaner cleaner) {
        return new MemorySessionImpl(new ConfinedSessionState(thread, cleaner));
    }

    public final static MemorySessionImpl createShared(Cleaner cleaner) {
        return new MemorySessionImpl(new SharedSessionState(cleaner));
    }

    public final static MemorySessionImpl createImplicit(Object ref) {
        return new MemorySessionImpl(new SharedSessionState.OfImplicit()) {
            final Object o = ref;

            @Override
            public boolean isCloseable() {
                return false;
            }
        };
    }

    public abstract static class State {

        static final int OPEN = 0;
        static final int CLOSING = -1;
        static final int CLOSED = -2;

        int state = OPEN;

        static final VarHandle STATE;

        static {
            try {
                STATE = MethodHandles.lookup().findVarHandle(State.class, "state", int.class);
            } catch (Throwable ex) {
                throw new ExceptionInInitializerError(ex);
            }
        }

        static final int MAX_FORKS = Integer.MAX_VALUE;

        final Cleaner.Cleanable cleanable;
        final ResourceList resourceList;

        State(ResourceList resourceList, Cleaner cleaner) {
            this.resourceList = resourceList;
            cleanable = cleaner != null ?
                    cleaner.register(this, resourceList) :
                    null;
        }

        /**
         * Closes this session, executing any cleanup action (where provided).
         *
         * @throws IllegalStateException if this session is already closed or if this is
         *                               a confined session and this method is called outside of the owner thread.
         */
        public final void close() {
            try {
                justClose();
                if (cleanable != null) {
                    cleanable.clean();
                } else {
                    resourceList.cleanup();
                }
            } finally {
                Reference.reachabilityFence(this);
            }
        }

        abstract void justClose();

        public final void checkValidStateWrapException() {
            try {
                checkValidState();
            } catch (ScopedMemoryAccess.ScopedAccessError ex) {
                throw ex.newRuntimeException();
            }
        }

        public boolean isImplicit() {
            return false;
        }

        public abstract boolean isAlive();

        public abstract Thread ownerThread();

        @ForceInline
        public void checkValidState() {
            if (ownerThread() != null && ownerThread() != Thread.currentThread()) {
                throw WRONG_THREAD;
            } else if (state < OPEN) {
                throw ALREADY_CLOSED;
            }
        }

        public abstract void acquire();

        public abstract void release();

        static IllegalStateException tooManyAcquires() {
            return new IllegalStateException("Session acquire limit exceeded");
        }

        static IllegalStateException alreadyAcquired(int acquires) {
            return new IllegalStateException(String.format("Session is acquired by %d clients", acquires));
        }

        static IllegalStateException alreadyClosed() {
            return new IllegalStateException("Already closed");
        }

        static WrongThreadException wrongThread() {
            return new WrongThreadException("Attempted access outside owning thread");
        }

        static final ScopedMemoryAccess.ScopedAccessError ALREADY_CLOSED = new ScopedMemoryAccess.ScopedAccessError(State::alreadyClosed);

        static final ScopedMemoryAccess.ScopedAccessError WRONG_THREAD = new ScopedMemoryAccess.ScopedAccessError(State::wrongThread);

        /**
         * A list of all cleanup actions associated with a memory session. Cleanup actions are modelled as instances
         * of the {@link ResourceCleanup} class, and, together, form a linked list. Depending on whether a session
         * is shared or confined, different implementations of this class will be used, see {@link ConfinedSessionState.ConfinedList}
         * and {@link SharedSessionState.SharedList}.
         */
        public abstract static class ResourceList implements Runnable {
            ResourceCleanup fst;

            public abstract void add(ResourceCleanup cleanup);

            abstract void cleanup();

            public final void run() {
                cleanup(); // cleaner interop
            }

            static void cleanup(ResourceCleanup first) {
                ResourceCleanup current = first;
                while (current != null) {
                    current.cleanup();
                    current = current.next;
                }
            }
        }

        public abstract static class ResourceCleanup implements Runnable {
            ResourceCleanup next;

            public abstract void cleanup();

            public final void run() {
                cleanup();
            }

            static final ResourceCleanup CLOSED_LIST = new ResourceCleanup() {
                @Override
                public void cleanup() {
                    throw new IllegalStateException("This resource list has already been closed!");
                }
            };

            public static ResourceCleanup ofRunnable(Runnable cleanupAction) {
                return cleanupAction instanceof ResourceCleanup ?
                        (ResourceCleanup)cleanupAction :
                        new ResourceCleanup() {
                            @Override
                            public void cleanup() {
                                cleanupAction.run();
                            }
                        };
            }
        }
    }
}
