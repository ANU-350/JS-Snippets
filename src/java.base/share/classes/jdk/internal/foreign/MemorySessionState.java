package jdk.internal.foreign;

import jdk.internal.misc.ScopedMemoryAccess;

import java.lang.foreign.MemorySession;
import java.lang.invoke.MethodHandles;
import java.lang.invoke.VarHandle;
import java.lang.ref.Cleaner;
import java.lang.ref.Reference;

public abstract class MemorySessionState {

    static final int OPEN = 0;
    static final int CLOSING = -1;
    static final int CLOSED = -2;

    int state = OPEN;

    static final VarHandle STATE;

    static {
        try {
            STATE = MethodHandles.lookup().findVarHandle(MemorySessionState.class, "state", int.class);
        } catch (Throwable ex) {
            throw new ExceptionInInitializerError(ex);
        }
    }

    static final int MAX_FORKS = Integer.MAX_VALUE;

    final Cleaner.Cleanable cleanable;
    public final ResourceList resourceList;

    MemorySessionState(ResourceList resourceList, Cleaner cleaner) {
        this.resourceList = resourceList;
        cleanable = cleaner != null ?
                cleaner.register(this, resourceList) :
                null;
    }

    /**
     * Closes this session, executing any cleanup action (where provided).
     *
     * @throws IllegalStateException if this session is already closed or if this is
     *                               a confined session and this method is called outside of the owner thread.
     */
    public final void close() {
        try {
            justClose();
            if (cleanable != null) {
                cleanable.clean();
            } else {
                resourceList.cleanup();
            }
        } finally {
            Reference.reachabilityFence(this);
        }
    }

    abstract void justClose();

    public final void checkValidStateWrapException() {
        try {
            checkValidState();
        } catch (ScopedMemoryAccess.ScopedAccessError ex) {
            throw new IllegalStateException("Already closed!");
        }
    }

    public abstract boolean isAlive();

    public abstract Thread ownerThread();

    public abstract void checkValidState();

    public abstract void acquire();

    public abstract void release();

    /**
     * A list of all cleanup actions associated with a memory session. Cleanup actions are modelled as instances
     * of the {@link ResourceList.ResourceCleanup} class, and, together, form a linked list. Depending on whether a session
     * is shared or confined, different implementations of this class will be used, see {@link ConfinedSessionState.ConfinedResourceList}
     * and {@link SharedSessionState.SharedResourceList}.
     */
    public abstract static class ResourceList implements Runnable {
        ResourceList.ResourceCleanup fst;

        public abstract void add(ResourceList.ResourceCleanup cleanup);

        abstract void cleanup();

        public final void run() {
            cleanup(); // cleaner interop
        }

        static void cleanup(ResourceList.ResourceCleanup first) {
            ResourceList.ResourceCleanup current = first;
            while (current != null) {
                current.cleanup();
                current = current.next;
            }
        }

        public abstract static class ResourceCleanup implements Runnable {
            ResourceList.ResourceCleanup next;

            public abstract void cleanup();

            public final void run() {
                cleanup();
            }

            static final ResourceList.ResourceCleanup CLOSED_LIST = new ResourceList.ResourceCleanup() {
                @Override
                public void cleanup() {
                    throw new IllegalStateException("This resource list has already been closed!");
                }
            };

            public static ResourceList.ResourceCleanup ofRunnable(Runnable cleanupAction) {
                return cleanupAction instanceof ResourceCleanup ?
                        (ResourceCleanup)cleanupAction :
                        new ResourceList.ResourceCleanup() {
                    @Override
                    public void cleanup() {
                        cleanupAction.run();
                    }
                };
            }
        }

    }
}
