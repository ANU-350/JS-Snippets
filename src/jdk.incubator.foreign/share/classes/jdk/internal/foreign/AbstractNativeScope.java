/*
 * copyright (c) 2020 oracle and/or its affiliates. all rights reserved.
 * do not alter or remove copyright notices or this file header.
 *
 * this code is free software; you can redistribute it and/or modify it
 * under the terms of the gnu general public license version 2 only, as
 * published by the free software foundation.  oracle designates this
 * particular file as subject to the "classpath" exception as provided
 * by oracle in the license file that accompanied this code.
 *
 * this code is distributed in the hope that it will be useful, but without
 * any warranty; without even the implied warranty of merchantability or
 * fitness for a particular purpose.  see the gnu general public license
 * version 2 for more details (a copy is included in the license file that
 * accompanied this code).
 *
 * you should have received a copy of the gnu general public license version
 * 2 along with this work; if not, write to the free software foundation,
 * inc., 51 franklin st, fifth floor, boston, ma 02110-1301 usa.
 *
 * please contact oracle, 500 oracle parkway, redwood shores, ca 94065 usa
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */
package jdk.internal.foreign;

import jdk.incubator.foreign.MemorySegment;
import jdk.incubator.foreign.NativeScope;

import java.util.ArrayList;
import java.util.List;
import java.util.OptionalLong;

public abstract class AbstractNativeScope implements NativeScope {

    private final List<MemorySegment> segments = new ArrayList<>();
    private final Thread ownerThread;

    private static final int SCOPE_MASK = MemorySegment.READ | MemorySegment.WRITE; // no terminal operations allowed

    AbstractNativeScope(Thread ownerThread) {
        this.ownerThread = ownerThread;
    }

    @Override
    public Thread ownerThread() {
        return ownerThread;
    }

    @Override
    public void close() {
        segments.forEach(MemorySegment::close);
    }

    void checkOwnerThread() {
        if (Thread.currentThread() != ownerThread()) {
            throw new IllegalStateException("Attempt to access scope from different thread");
        }
    }

    MemorySegment newSegment(long size, long align) {
        MemorySegment segment = MemorySegment.allocateNative(size, align);
        segments.add(segment);
        return segment;
    }

    MemorySegment newSegment(long size) {
        return newSegment(size, size);
    }

    public void register(MemorySegment segment) {
        segments.add(segment);
    }

    public static class UnboundedNativeScope extends AbstractNativeScope {

        private static final long BLOCK_SIZE = 4 * 1024;
        private static final long MAX_ALLOC_SIZE = BLOCK_SIZE / 2;

        private MemorySegment segment;
        private long sp = 0L;
        private long size = 0L;

        @Override
        public OptionalLong byteSize() {
            return OptionalLong.empty();
        }

        @Override
        public long allocatedBytes() {
            return size;
        }

        public UnboundedNativeScope() {
            super(Thread.currentThread());
            this.segment = newSegment(BLOCK_SIZE);
        }

        @Override
        public MemorySegment allocate(long bytesSize, long bytesAlignment) {
            checkOwnerThread();
            if (bytesSize > MAX_ALLOC_SIZE) {
                MemorySegment segment = newSegment(bytesSize, bytesAlignment);
                return segment.withAccessModes(SCOPE_MASK);
            }
            for (int i = 0; i < 2; i++) {
                long min = segment.address().toRawLongValue();
                long start = Utils.alignUp(min + sp, bytesAlignment) - min;
                try {
                    MemorySegment slice = segment.asSlice(start, bytesSize)
                            .withAccessModes(SCOPE_MASK);
                    sp = start + bytesSize;
                    size += Utils.alignUp(bytesSize, bytesAlignment);
                    return slice;
                } catch (IndexOutOfBoundsException ex) {
                    sp = 0L;
                    segment = newSegment(BLOCK_SIZE, 1L);
                }
            }
            throw new AssertionError("Cannot get here!");
        }
    }

    public static class BoundedNativeScope extends AbstractNativeScope {
        private final MemorySegment segment;
        private long sp = 0L;

        @Override
        public OptionalLong byteSize() {
            return OptionalLong.of(segment.byteSize());
        }

        @Override
        public long allocatedBytes() {
            return sp;
        }

        public BoundedNativeScope(long size) {
            super(Thread.currentThread());
            this.segment = newSegment(size, 1);
        }

        @Override
        public MemorySegment allocate(long bytesSize, long bytesAlignment) {
            checkOwnerThread();
            long min = segment.address().toRawLongValue();
            long start = Utils.alignUp(min + sp, bytesAlignment) - min;
            try {
                MemorySegment slice = segment.asSlice(start, bytesSize)
                        .withAccessModes(SCOPE_MASK);
                sp = start + bytesSize;
                return slice;
            } catch (IndexOutOfBoundsException ex) {
                throw new OutOfMemoryError("Not enough space left to allocate");
            }
        }
    }
}
