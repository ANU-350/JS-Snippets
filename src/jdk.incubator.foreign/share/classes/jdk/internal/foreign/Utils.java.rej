diff a/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/Utils.java b/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/Utils.java	(rejected hunks)
@@ -28,9 +28,12 @@ package jdk.internal.foreign;
 
 import jdk.incubator.foreign.MemoryAddress;
 import jdk.incubator.foreign.MemoryHandles;
+import jdk.incubator.foreign.MemoryLayout;
 import jdk.incubator.foreign.MemorySegment;
+import jdk.incubator.foreign.ValueLayout;
 import jdk.internal.access.foreign.MemorySegmentProxy;
 import jdk.internal.misc.VM;
+import sun.invoke.util.Wrapper;
 
 import java.lang.invoke.MethodHandle;
 import java.lang.invoke.MethodHandles;
