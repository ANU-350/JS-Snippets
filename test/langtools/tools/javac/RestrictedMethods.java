/*
 * @test /nodynamiccopyright/
 * @bug 8316971
 * @summary Smoke test for restricted method call warnings
 * @compile/fail/ref=RestrictedMethods.out -Xlint:restricted -Werror -XDrawDiagnostics --enable-preview --source ${jdk.version} RestrictedMethods.java
 * @compile -Werror --enable-preview --source ${jdk.version} RestrictedMethods.java
 */

import java.lang.foreign.MemorySegment;

class RestrictedMethods {

    MemorySegment warn = MemorySegment.NULL.reinterpret(10);
    @SuppressWarnings("restricted")
    MemorySegment suppressed = MemorySegment.NULL.reinterpret(10);

    void testWarn() {
        MemorySegment.NULL.reinterpret(10); // warning here
    }

    @SuppressWarnings("restricted")
    void testSuppressed() {
        MemorySegment.NULL.reinterpret(10); // no warning here
    }

    @SuppressWarnings("restricted")
    static class Nested {
        MemorySegment suppressedNested = MemorySegment.NULL.reinterpret(10);

        void testSuppressedNested() {
            MemorySegment.NULL.reinterpret(10); // no warning here
        }
    }
}
