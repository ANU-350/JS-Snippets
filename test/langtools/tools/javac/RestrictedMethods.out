RestrictedMethods.java:13:44: compiler.warn.restricted.method.call: reinterpret(long)
RestrictedMethods.java:18:27: compiler.warn.restricted.method.call: reinterpret(long)
- compiler.err.warnings.and.werror
- compiler.note.preview.filename: RestrictedMethods.java, DEFAULT
- compiler.note.preview.recompile
1 error
2 warnings
